File: 03_pattern_matching/geom.erl
-module(geom).
-export([area/3]).
-include_lib("eunit/include/eunit.hrl").

-spec(area(atom(), number(), number()) -> number()).
area(rectangle, Width, Height) when Width >= 0, Height >= 0 ->
  Width * Height;
area(triangle, Base, Height) when Base >= 0, Height >= 0 ->
  Base * Height / 2.0;
area(ellipse, Width, Height) when Width >= 0, Height >= 0 ->
  math:pi() * Width * Height;
area(_, _Width, _Height) -> 0.


area_test_() -> [
  ?_assert( area(rectangle, 3,4)   =:= 12  ),
  ?_assert( area(rectangle, 0,4)   =:= 0   ),
  ?_assert( area(rectangle, 1.5,4) =:= 6.0 ),
  ?_assert( area(triangle, 3,5)    =:= 7.5 ),
  ?_assert( area(ellipse, 2,4) > 25.132 andalso area(ellipse, 2,4) < 25.133),
  ?_assert( area(rectangle, -1,1)  =:= 0 ),
  ?_assert( area(pentagon, 1, 1)   =:= 0 )
].

File: ex04_logic_recursion/lib/geom.ex
defmodule Geom do
  @moduledoc """
    Calculates the area of geometric shapes.
    """

  @doc "area({shape, width, height}) -> the area of a shape with the given width and height.
  supported shapes = rectangle, triangle, ellipse"
  def area({shape, width, height}) do
    area(shape, width, height)
  end

  defp area(shape, width, height) when width > 0 and height > 0 do
    case shape do
      :rectangle -> (width * height)
      :triangle -> (width * height) / 2.0
      :ellipse -> :math.pi * width * height
      _ -> 0
    end
  end

  defp area(_shape, _width, _height) do
    0
  end
end
